sled-agent/src/services.rs

```diff
@@ -2878,211 +2927,224 @@ impl ServiceManager {
         let time_is_synchronized =
             match self.timesync_get_locked(&existing_zones).await {
                 // Time is synchronized
                 Ok(TimeSync { sync: true, .. }) => true,
                 // Time is not synchronized, or we can't check
                 _ => false,
             };
-        for zone in zones_to_be_added.clone() {
-            if zone_requires_timesync(&zone.zone_type) && !time_is_synchronized
-            {
-                return Err(Error::TimeNotSynchronized);
-            }
+
+        // Concurrently boot all new zones
+        let StartZonesResult { zones, errors } = self
+            .start_omicron_zones(
+                zones_to_be_added,
+                time_is_synchronized,
+                &all_u2_pools,
+            )
+            .await?;
+
+        // Add the new zones to our tracked zone set
+        existing_zones.extend(
+            zones.into_iter().map(|zone| (zone.name().to_string(), zone)),
+        );
+
+        // If any zones failed to start, exit with an error
+        if !errors.is_empty() {
+            return Err(Error::ZoneEnsure { errors });
         }
+        Ok(())
+    }
 
-        // Destroy zones that should not be running
-        for zone in zones_to_be_removed {
-            let expected_zone_name = zone.zone_name();
-            if let Some(mut zone) = existing_zones.remove(&expected_zone_name) {
-                debug!(
+    // Attempts to take a zone bundle and remove a zone.
+    //
+    // Logs, but does not return an error on failure.
+    async fn zone_bundle_and_try_remove(
+        &self,
+        existing_zones: &mut MutexGuard<'_, ZoneMap>,
+        zone: &OmicronZoneConfig,
+    ) {
+        let log = &self.inner.log;
+        let expected_zone_name = zone.zone_name();
+        if let Some(mut zone) = existing_zones.remove(&expected_zone_name) {
+            debug!(
+                log,
+                "removing an existing zone";
+                "zone_name" => &expected_zone_name,
+            );
+            if let Err(e) = self
+                .inner
+                .zone_bundler
+                .create(&zone.runtime, ZoneBundleCause::UnexpectedZone)
+                .await
+            {
+                error!(
                     log,
-                    "removing an existing zone";
+                    "Failed to take bundle of unexpected zone";
                     "zone_name" => &expected_zone_name,
+                    "reason" => ?e,
                 );
-                if let Err(e) = self
-                    .inner
-                    .zone_bundler
-                    .create(&zone, ZoneBundleCause::UnexpectedZone)
-                    .await
+            }
+            if let Err(e) = zone.runtime.stop().await {
+                error!(log, "Failed to stop zone {}: {e}", zone.name());
+            }
+        } else {
+            warn!(log, "Expected to remove zone, but could not find it");
```

Comment:
Maybe add `expected_zone_name` to this log?

Tiny style nitpick: You could structure this as

```rust
let Some(mut zone) = existing_zones.remove(&expected_zone_name) else {
    warn!(...);
    return;
};
```

to shave an indentation level off the bulk of the method.

Reply:
Done and done